common/error.hpp

```diff
@@ -13,31 +13,42 @@
 //    limitations under the License.
 
 #include <string>
+#include <system_error>
 #include <type_traits>
 
 #ifndef MENDER_COMMON_ERROR_HPP
 #define MENDER_COMMON_ERROR_HPP
 
 namespace mender::common::error {
 
-template <typename ErrorCodeType>
+template <typename ErrorType>
 class Error {
 public:
-	static_assert(
-		std::is_enum<ErrorCodeType>::value, "Error requires an enum type for error codes");
-	ErrorCodeType error_code;
+	ErrorType error;
 	std::string message;
 
-	Error(const ErrorCodeType &ec, const std::string &msg) {
-		this->error_code = ec;
-		this->message = msg;
-	};
-	Error(const Error &e) {
-		this->error_code = e.error_code;
-		this->message = e.message;
-	};
+	Error() {
+	}
+	Error(const ErrorType &ec, const std::string &msg) :
+		error(ec),
+		message(msg) {
+	}
+	Error(const Error &e) :
+		error(e.error),
+		message(e.message) {
+	}
+
+	bool operator==(const Error &other) const {
+		return this->message == other.message && this->error == other.error;
+	}
+
+	operator bool() const {
+		return static_cast<bool>(this->error);
+	}
 };
 
+typedef Error<std::error_condition> StdError;
```

Comment:
What is the difference between this and `using StdError = Error<std:error_condition>`? I take it the above creates a new type while the `using` clause only gives a shorter name to the template instantiation?

Reply:
Good point. Using `using` (fun to say) is probably better here.